**Сергей, привет!**

Это опять я, Алексей Каргин, я буду проверять проект по **SQL** в рамках финального спринта. Как всегда для удобства я оставлю комментарии в отдельных ячейках Markdown с заголовком «Комментарий ревьюера». Пожалуйста, не перемещай, не изменяй и не удаляй их - так наше общение будет более продуктивным. Я буду использовать цветовую разметку:

<div class="alert alert-danger">
<b>Комментарий ревьюера</b> 
    
✍ Так выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера</b> 
    
📝 Так выделены небольшие замечания или рекомендации. Постарайся учесть эти комментарии в этом проекте.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера</b> 
    
👍 Так выделены все остальные комментарии, включая рекомендации, позитивные моменты или какие-то рассуждения.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комментарии — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Например, вот так.
</div>

---

<div class="alert alert-success">
<b>Обратная связь v.1</b> 
    
👋 Сергей, спасибо, что прислал проект! Ты хорошо придерживаешься оформления проекта, твои запросы легко читать, и они также хорошо оформлены. Мы можем также добавлять небольшие комментарии к подзапросам - так проще возвращаться к своим запросам и проектам через некоторое время.
    
<br>
<div class="text-danger">   
    
✍ Однако проект пока нельзя принять, так как один из пунктов задания был пропущен. Дополнительно посмотри, пожалуста, четвертый запрос - его нужно немного скорректировать.
       
</div>    
    
Мне кажется, что несмотря на все мои комментарии у тебя не составит труда внести необходимые изменения. Буду ждать новой версии проекта!

<br>    
С наилучшими пожеланиями, <br>
Алексей
</div>

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Привет, Алексей!
    
Скорректировал четвертый запрос и добавил просмотр таблиц перед обработкой запросов.=)
    
Спасибо за обратную связь!
</div>

---

<div class="alert alert-success">
<b>Обратная связь v.2</b> 
    
👋 Сергей, спасибо, что оперативно внес правки в четвертый запрос и добавил необходимое задание. Теперь все верно и проект может быть принят! Поздравляю с окончанием финального спринта и обучения на курсе Практикума 🎉
    
Как тимлид финального спринта я еще напишу с Слаке финальную обратную связь, но это будет где-то ближе к финальному созвону. Было интересно проверять твои проекты, спасибо!
    

<br>    
С наилучшими пожеланиями, <br>
Алексей
</div>


# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Описание данных**

Таблица `books`

Содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.

Таблица `authors`

Содержит данные об авторах:

* `author_id` — идентификатор автора;
* `author` — имя автора.

Таблица `publishers`

Содержит данные об издательствах:

* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

Таблица `ratings`

Содержит данные о пользовательских оценках книг:

* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

Таблица `reviews`

Содержит данные о пользовательских обзорах:

* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.

**План работы**

1. Загрузка данных
2. Исследование таблиц
 * Подсчет, сколько книг вышло после 1 января 2000 года;
 * Для каждой книги подсчет количества обзоров и средней оценки;
 * Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (без брошюр);
 * Определение автора с самой высокой средней оценкой книг с 50 и более оценками;
 * Подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок.
3. Выводы

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 Хорошо, что в начале проекта есть краткое описание, а также данные о исходных таблицах. Также мы можем добавить схему базы данных, чтобы понимать, как таблицы связаны между собой.
</div>

## Загрузка данных

Импортируем библиотеки и подготовим соединение с базой для запросов.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student',
             'pwd': 'Sdf4$2;d-d30pp',  
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
             'port': 6432, 
             'db': 'data-analyst-final-project-db'} 

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Перейдем к исследованию таблиц и к ответам на поставленные вопросы.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b>
    
👍 Все готово для подключения к базе данных! 

</div>

## Исследование таблиц

<div class="alert alert-danger">
<b>Комментарий ревьюера v.1</b> 
    
✍ Сергей, согласно заданию к проекту, нам нужно выполнить следующий пункт "Исследуйте таблицы — выведите первые строки". Пожалуйста, проведи дополнительные исследования в этой части проекта. Нам важно убедиться, что подключение работает и все данные на месте и они надлежащего качества.
</div> 

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Вывел первые строки и проверил типы данных.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b>
    
👍 Отлично! Подключение к базе  данных работает, таблицы на месте. На экран были выведены первые строки из каждой таблицы, что позволило оценить их содержание, а также основная информация. Однако, обрати, пожалуйста, внимание, что в запросе мы загружаем все данные, поэтому перед их выгрузкой лучше оценить их объем. Почему это важно - если данных много, а их может быть много и они могут быть объемными, мы можем создать дополнительную нагрузку на окружение, что может быть не очень хорошо.
</div>

Проверим установленное соединение и посмтрим первые строки таблиц, чтобы убедиться, что все данные на месте. Также посмотрим типы данных по столбцам на соответствие указанным типам в схеме данных.

In [3]:
query = '''SELECT *
           FROM books
        '''

In [4]:
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


В таблице с книгами видим, что все данные присутствуют, типы данных соответствуют заявленным.

In [6]:
query = '''SELECT *
           FROM authors
        '''

In [7]:
authors = pd.io.sql.read_sql(query, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


В таблице с авторами также видим, что все данные присутствуют, типы данных соответствуют заявленным.

In [9]:
query = '''SELECT *
           FROM publishers
        '''

In [10]:
publishers = pd.io.sql.read_sql(query, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [11]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


В таблице с издательствами аналогично: все данные присутствуют, типы данных соответствуют заявленным.

In [12]:
query = '''SELECT *
           FROM ratings
        '''

In [13]:
ratings = pd.io.sql.read_sql(query, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [14]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


В таблице с оценками видим, что все данные присутствуют, типы данных соответствуют заявленным.

In [15]:
query = '''SELECT *
           FROM reviews
        '''

In [16]:
reviews = pd.io.sql.read_sql(query, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [17]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


В таблице с отзывами также все данные присутствуют, типы данных соответствуют заявленным.
Данные корректны во всех таблицах, перейдем к выполнению запросов.

### Подсчет, сколько книг вышло после 1 января 2000 года

Произведем запрос SQL и выведем количество книг.

In [18]:
query = '''SELECT COUNT(book_id) AS cnt_books
           FROM books
           WHERE publication_date > '2000-01-01'
        '''

In [19]:
cnt_books = pd.io.sql.read_sql(query, con = engine) 
cnt_books

cnt_books
0        819

После первого января 2000 года выпущено 819 книг.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 1:</b> Тут все верно! 
</div>

### Для каждой книги подсчет количества обзоров и средней оценки;

Сделаем SQL запрос и выведем таблицу с количеством обзоров и средней оценкой по каждой книге.

In [20]:
query = '''SELECT
             books.book_id AS book_id,             
             books.title AS title,
             rev_cnt.cnt_reviews AS cnt_reviews,
             avg_rat.avg_rating AS avg_rating             
           FROM
             books
             LEFT JOIN (SELECT
                           book_id,
                           COUNT(review_id) AS cnt_reviews             
                         FROM
                           reviews
                         GROUP BY
                           book_id) AS rev_cnt ON rev_cnt.book_id = books.book_id
             LEFT JOIN (SELECT
                           book_id,
                           AVG(rating) AS avg_rating             
                         FROM
                           ratings
                         GROUP BY
                           book_id) AS avg_rat ON avg_rat.book_id = books.book_id
           
        '''

In [21]:
books_reviews_rating = pd.io.sql.read_sql(query, con = engine)
books_reviews_rating

book_id                                              title  cnt_reviews  \
0        652          The Body in the Library (Miss Marple  #3)          2.0   
1        273                                          Galápagos          2.0   
2         51                           A Tree Grows in Brooklyn          5.0   
3        951  Undaunted Courage: The Pioneering First Missio...          2.0   
4        839                                        The Prophet          4.0   
..       ...                                                ...          ...   
995       64                                Alice in Wonderland          4.0   
996       55          A Woman of Substance (Emma Harte Saga #1)          2.0   
997      148                                          Christine          3.0   
998      790  The Magicians' Guild (Black Magician Trilogy  #1)          2.0   
999      828                           The Plot Against America          2.0   

     avg_rating  
0      4.500000  
1      4.500000  
2      4.250000  
3      4.000000  
4      4.285714  
..          ...  
995    4.230769  
996    5.000000  
997    3.428571  
998    3.500000  
999    3.000000  

[1000 rows x 4 columns]

Видим, что у каждой книги есть по несколько отзывов, средняя оценка разнится.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 2:</b> Все корректно, здорово, что в начале ты последовательно подсоединил подсчитанные количество обзоров и средний рейтинг. Использование левого объединение позволило не упустить ни одну книгу. Все верно!
</div>

### Определение издательства, которое выпустило наибольшее число книг толще 50 страниц (без брошюр)

Сделаем SQL запрос и найдем издательство, которое выпустило наибольшее количество книг тольще 50 страниц.

In [22]:
query = '''SELECT
             publishers.publisher AS publisher,
             COUNT(books.book_id) AS cnt_books
           FROM
             publishers
             LEFT JOIN books ON books.publisher_id = publishers.publisher_id
           WHERE
             books.num_pages > 50
           GROUP BY
             publisher
           ORDER BY
             cnt_books DESC
           LIMIT
             1
        '''

In [23]:
top_publisher = pd.io.sql.read_sql(query, con = engine)
top_publisher

publisher  cnt_books
0  Penguin Books         42

Издатель Penguin Books выпустил наибольшее количество книг - 42 штуки.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 3:</b> Тут тоже все корректно.
</div>

### Определение автора с самой высокой средней оценкой книг с 50 и более оценками

Сделаем SQL запрос и и определим топ автора.

In [32]:
query = '''SELECT
             authors.author AS author,
             AVG(ratings_50.avg_rating) AS avg_rating,
             COUNT(books.book_id) AS cnt_books
           FROM
             books
             INNER JOIN (SELECT
                          book_id AS book_id,
                          AVG(rating) AS avg_rating                          
                        FROM
                          ratings
                        GROUP BY
                          book_id                          
                        HAVING
                          COUNT(rating_id) > 50) AS ratings_50 ON ratings_50.book_id = books.book_id
             LEFT JOIN authors ON authors.author_id = books.author_id
           GROUP BY
             author
           ORDER BY
             avg_rating DESC
           LIMIT 1
        '''

In [33]:
top_author = pd.io.sql.read_sql(query, con = engine)
top_author

author  avg_rating  cnt_books
0  J.K. Rowling/Mary GrandPré    4.283844          4

Автор с самой высокой средней оценкой с 50 и более оценками является J.K. Rowling/Mary GrandPré. Средний рейтинг составлен на основе оценки 4-х книг.

<div class="alert alert-danger">
<b>Комментарий ревьюера v.1</b> 
    
📝 <b>Запрос 4:</b> А вот этот запрос нужно немного скорректировать. Смотри, по заданию, нам в начале нужно выбрать книги, у которых 50 или более оценок и уже используя только эти книги подсчитать средний рейтинг для каждой книги и агрегировать данные по автору. Сейчас мы выбрали авторов, у которых количество оценок для всех книг 50 или более. Пожалуйста, скорректируй этот запрос. Также будет полезно вывести количество книг, которые вошли в подсчет (опционально).
</div> 

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Понял, поправил, видимо я не так воспринял задание в этом пункте, мне почему-то показалось, что автора у которого все книги заработали более 50 оценок. Переделал, добавил подсчет книг.=) Надеюсь, что все верно.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 <b>Запрос 4:</b> Спасибо, что объяснил почему сделал запрос таким образом. Теперь все корректно! Спасибо, что проработал этот момент.
</div>

### Подсчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

Сделаем SQL запрос и подсчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [26]:
query = '''SELECT
             AVG(Sub.cnt_reviews) AS avg_reviews
           FROM
             (SELECT
                username,
                COUNT(review_id) AS cnt_reviews
              FROM
                reviews
              WHERE
                username IN (SELECT
                                 username
                               FROM
                                 ratings
                               GROUP BY
                                 username
                               HAVING
                                 COUNT(rating_id) >50)
              GROUP BY
                username) AS Sub
            
        '''

In [27]:
avg_count_reviews = pd.io.sql.read_sql(query, con = engine)
avg_count_reviews

avg_reviews
0    24.333333

Среднее количество обзоров от пользователей, поставивших более 50 оценок - 24.33. Это говорит о том, что активные читатели оставляют отзывы не на все прочитанные книги. Отзывов как миниму в половину меньше, чем оценок.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 <b>Запрос 5:</b> А здесь все верно!
</div>

## Выводы

После проведения исследования таблиц с помощью SQL-запросов выяснили следующее:

1. Из 1000 книг, 819 - это книги, выпущенные после 01 января 2000 года. Т.е. преимущественно книги современного авторства превалируют. Можно данную информацию выделить на входной страницедля привлечения аудитории по тэгам.
2. У каждой книги есть как минимум несколько отзывов и средняя оценка, что позволяет в данном списке выделить наиболее интересные и понравившиеся книги для новых читателей, сделать отбор и рекомендации.
3. Издатель Penguin Books выпустил наибольшее количество книг - 42 штуки. Его активность говорит о том, что стоит изучить ассортимент выпускаемых книг, возможно, что выпускаемые им книги пользуются популярностью и их можно выделить, как отдельный контент с ссылкой на данное издательство на входной странице.
4. Автор с самой высокой средней оценкой с 50 и более оценками является J.K. Rowling/Mary GrandPré. Оценка составлена на средней оценке 4 книг этого автора, возможно это Гарри Поттер. Стоит это отметить на входной странице как топ автор и его книги выделить в отдельный контент с ссылкой на входной странице.
5. Среднее количество обзоров от пользователей, поставивших более 50 оценок - 24.33. Это говорит о том, что активные читатели оставляют отзывы не на все прочитанные книги. Отзывов как миниму в половину меньше, чем оценок. Возможно, что следует добавить пояснение для читателей при оценке, что его отзыв поможет другим читателям в выборе книги, это повысит частоту написания отзывов.

<div class="alert alert-success">
<b>Комментарий ревьюера v.1</b> 
    
👍 Отлично, что в проекте есть небольшое резюме по результатам запросов! Не забудь, пожалуйста, проверить его после работы с комментариями.

</div>

<div class="alert alert-info"> 
<b>Комментарий студента</b> 
    
Подкорректировал=)
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера v.2</b> 
    
👍 Спасибо!
</div>